[IR][Instructions] Add `CallBase::getCalledFunctionName` helper

tsymalla opened the pull request:
There's a lot of uses of `getCalledFunction()->getName()`, without ever checking if `getCalledFunction()` returns a nullptr. Add a new helper that returns an `std::optional<StringRef>`, that can be used to

- Avoid method chaining
- Have a more safe shortcut to get the callee name.

tsymalla committed:
Add CallBase::getCalledFunctionName

There's a lot of uses of `getCalledFunction()->getName()`, without ever
checking if `getCalledFunction()` returns a nullptr. Add a new helper
that returns an `std::optional<StringRef>`, that can be used to

- Avoid method chaining
- Have a more safe shortcut to get the callee name.

nikic commented on llvm/include/llvm/IR/InstrTypes.h:
```diff
@@ -1345,6 +1345,16 @@ class CallBase : public Instruction {
+    if (F)
+      return F->getName();
+
+    return std::nullopt;
```
Can we use an empty StringRef as the sentinel instead?

tsymalla replied:
Technically, yes, but I think `std::optional` is a better fit, since using `operator*` on the return value would fail if no value is present.

arsenm replied:
Which is the kind of failure you are trying to avoid needing to worry about?

nikic reviewed (commented):
If you're introducing a new helper API, please also update existing code that can make use of it.

tsymalla committed:
Some refactorings to make use of the new helper.

`LibCallsShrinkWrap::perform`: This uses the potentially `nullptr` value
inside the loop in a debug statement before asserting on it. Move the
assertion outside `perform(CallInst &)`.

Some other arbitrary uses to replace `getCalledFunction()->getName()`.

nikic reviewed (commented):
From a quick git grep "getCalledFunction()->getName()", it looks like there are a lot more places that could use this.

tsymalla commented:
> From a quick git grep "getCalledFunction()->getName()", it looks like there are a lot more places that could use this.

Yep. I wanted to see if the change is generally accepted at first, before changing all the code. I'll follow up with the remainder.

tsymalla committed:
Use `getCalledFunctionName()` throughout the codebase.

tsymalla committed:
Fix invalid whitespace character

tsymalla committed:
clang-format

tsymalla-AMD commented:
Sorry for the massive update. All callsites of `getCalledFunction()->getName()` should now be updated plus some extra changes.
